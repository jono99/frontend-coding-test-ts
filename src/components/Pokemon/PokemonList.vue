<template>
    <div class="mb-5">
        <select
            id="pokemon"
            v-model="selectedValue"
            name="pokemon"
            class="mt-1.5 w-1/4 rounded-lg border-gray-300 text-gray-700 sm:text-sm"
            v-on:change="onChange"
        >
            <option disabled value="" class="pokemon-item">
                Select a pokemon
            </option>
            <option
                v-for="pokemon in listPokemon"
                v-bind:key="pokemon.url"
                v-bind:value="pokemon.url"
                class="pokemon-item"
            >
                {{ capitalizeFirstLetter(pokemon.name) }}
            </option>
        </select>
    </div>
</template>

<script setup lang="ts">
import axios from 'axios';
import { defineEmits, PropType, ref, onBeforeMount } from 'vue';
import { useToast } from 'vue-toastification';

interface Pokemon {
    name: string;
    url: string;
}

const emits = defineEmits(['update']);
const toast = useToast();

const props = defineProps({
    selected: {
        type: String,
        default: '',
        required: true,
    },
});

const listPokemon = ref<Pokemon[]>([]);
const selectedValue = ref(props.selected);

const capitalizeFirstLetter = (string: string) =>
    string.charAt(0).toUpperCase() + string.slice(1);

const onChange = () => {
    emits('update', selectedValue.value);
};

onBeforeMount(async () => {
    await axios
        .get('https://pokeapi.co/api/v2/pokemon')
        .then((result) => {
            listPokemon.value = result.data.results;
            toast.success('Pokemon list loaded');
        })
        .catch((error) => {
            toast.error('Error loading list of Pokemon', error);
        });
});
</script>
