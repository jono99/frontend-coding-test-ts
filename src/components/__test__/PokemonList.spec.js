import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';
import axios from 'axios';
import PokemonList from '../Pokemon/PokemonList.vue';

vi.mock('axios');

describe('Pokemon Service', () => {
    beforeEach(() => {
        axios.get.mockReset();
    });

    describe('fetchPokemon', () => {
        test('make a GET request to fetch Pokemon', async () => {
            const pokemonMock = [
                {
                    id: 1,
                    name: 'bulbasaur',
                    url: 'https://pokeapi.co/api/v2/pokemon/1',
                },
                {
                    id: 4,
                    name: 'charmander',
                    url: 'https://pokeapi.co/api/v2/pokemon/4',
                },
                {
                    id: 7,
                    name: 'squirtle',
                    url: 'https://pokeapi.co/api/v2/pokemon/7',
                },
            ];
            axios.get.mockResolvedValue({
                data: pokemonMock,
            });

            const pokemon = await axios.get(
                'https://pokeapi.co/api/v2/pokemon',
            );

            expect(axios.get).toHaveBeenCalledWith(
                'https://pokeapi.co/api/v2/pokemon',
            );
            expect(pokemon).toStrictEqual(pokemonMock);

            // const wrapper = mount(PokemonList);
            // await wrapper.vm.$nextTick();

            // expect(wrapper.html()).toMatchSnapshot();
        });
    });
});
