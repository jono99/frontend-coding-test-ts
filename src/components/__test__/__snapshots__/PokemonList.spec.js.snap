// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pokemon Service > fetchPokemon > make a GET request to fetch Pokemon 1`] = `
"<div class="mb-5"><select id="pokemon" name="pokemon" class="mt-1.5 w-1/4 rounded-lg border-gray-300 text-gray-700 sm:text-sm">
    <option disabled="" class="pokemon-item" value=""> Select a pokemon </option>
  </select></div>"
`;

exports[`PokemonList > renders 1`] = `
"<div class="mb-5"><select id="pokemon" name="pokemon" class="mt-1.5 w-1/4 rounded-lg border-gray-300 text-gray-700 sm:text-sm">
    <option disabled="" class="pokemon-item" value=""> Select a pokemon </option>
  </select></div>"
`;
